docs/bundles.rst

```diff
@@ -0,0 +1,137 @@
+Bundles
+=========================
+
+The Blockstore does not parse or process files other than ``bundle.json``.
+However, some conventions should be followed by authoring clients for interoperability:
+
+* Files in a Bundle can be organized freely by authoring clients, including in directories.
+* A **bundle.json** file must be added to the root of the Bundle by authoring clients. This file specifies the
+  shareable components, bundle dependencies, assets and other metadata. See below for details.
+* OLX files must use the **.olx** extension. Files with a different extension may be ignored or not treated as
+  OLX by plugins and other clients. Files of any other type should not use this extension.
+
+Paths
+-------------------------
+
+Files can be referenced with paths which are relative references as specified in section 4.2 of
+`RFC 3986 <http://www.ietf.org/rfc/rfc3986.txt>`_.
+
+XBlock Runtimes are expected to provide a service which allows reading a file given a URI or for converting a URI into
+an internet accessible URL that can be passed to client applications.
+
+.. code-block::
+
+  file = blockstore_service.open(path_of_file_1)
+
+  url_1 = blockstore_service.url_for_path(path_of_directory_1)
+  url_2 = blockstore_service.url_for_path(path_of_file_2)
+
+For directories or files in the same Bundle as the XBlock, the URI must be an absolute-path reference
+which consists of a single slash character followed by the full path of the directory or file from the root
+of the Bundle.
+
+.. code-block::
+
+  /bundle.json
+  /mcqs/mcq1.olx
+  /b/c/icon.png
+  /d/README
+  /e/f/
+
+For files in other Bundles, the URI must be a network-path reference which begins with two slash characters followed
```

Comment:
One problem with this is that authors sometimes use `//example.com/images/image.png` in their content, where `//` gets resolved by the browser to be either `http` or `https`. So we might occasionally get conflicts.

Same issue applies to URLs like `/images/blah.png` - is that intending to refer to a path on the Open edX server, or to the "images" folder in the root of the current bundle?

I'm almost thinking we should use a prefix like `bundle:/image.png`. There's a lot to be said for having no ambiguity.

Also, did we consider @ormsbee's suggestion of "mounting" external bundles on virtual paths within the bundle? I think that was kind of nice. (e.g. `bundle:/deps/otherbundle/image.png`)

Reply:
I am fully onboard with having the scheme specified in the URI. Saving a few characters for convenience but adding complexity+ambiguity seems like the less desirable option to me.

> Also, did we consider @ormsbee's suggestion of "mounting" external bundles on virtual paths within the bundle? I think that was kind of nice. (e.g. bundle:/deps/otherbundle/image.png)

There was a conversation above: https://github.com/open-craft/blockstore/pull/5#discussion_r213142268. And this is already how the runtimes will have to work. The only difference from your example is that instead of `bundle:/debs/<bundle_alias>/<path_to_file>`, its going to be `bundle://<bundle_alias>/<path_to_file>/` where the `bundle_alias` is the URI authority/domain. Given that XBlocks do not deal with any filesystems (and so there are no libs which can benefit from virtual mounting), I do not see what advantages it provides compared to the points I have mentioned in the conversation thread above. Are there any I am missing?

Reply:
Ok, makes sense :) I don't think you're missing anything.